pkg/apis/batch/validation/validation.go

```diff
@@ -485,9 +524,36 @@ func validatePodTemplateUpdate(spec, oldSpec batch.JobSpec, fldPath *field.Path,
 }
 
 // ValidateJobStatusUpdate validates an update to a JobStatus and returns an ErrorList with any errors.
-func ValidateJobStatusUpdate(status, oldStatus batch.JobStatus) field.ErrorList {
+func ValidateJobStatusUpdate(job, oldJob *batch.Job) field.ErrorList {
 	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, validateJobStatus(&status, field.NewPath("status"))...)
+	statusFld := field.NewPath("status")
+	allErrs = append(allErrs, validateJobStatus(job, statusFld)...)
+
+	// A strengthened validation of the Job status transitions is needed since the
+	// Job managed-by label let's the Job object be controlled by external
+	// controllers. We want to make sure the transitions done by the external
+	// controllers meet the expectations of the clients of the Job API.
+	// In particular, we verify that a Job in terminal state (Failed or Complete)
+	// does not flip to a non-terminal state.
+	if utilfeature.DefaultFeatureGate.Enabled(features.JobManagedByLabel) {
+		for _, cType := range []batch.JobConditionType{batch.JobFailed, batch.JobComplete, batch.JobFailureTarget} {
```

Comment:
Its not clear to me why these validations are different from the ones in line 414.

Could we DRY this and maybe move these validations to a function?

Reply:
Oh, its an update.

Reply:
Yes, the validations here need the oldJob. For example, there is nothing wrong if a job has Failed=False, and Complete=False, but we don't want to transition to such state from Complete=True.